graph/src/data_source/common.rs

```diff
@@ -748,4 +841,283 @@ mod tests {
             CallArg::Subgraph(SubgraphArg::EntityParam(_))
         ));
     }
+
+    #[test]
+    fn test_struct_field_access_parsing() {
+        // Test struct field access with numeric indices
+        let arg = CallArg::from_str("event.params.myStruct.1").unwrap();
+        assert!(matches!(
+            arg,
+            CallArg::Ethereum(EthereumArg::ParamField(_, _))
+        ));
+        if let CallArg::Ethereum(EthereumArg::ParamField(param, field)) = arg {
+            assert_eq!(param.as_str(), "myStruct");
+            assert_eq!(field.as_str(), "1");
+        }
+
+        // Test struct field access with index 0
+        let arg = CallArg::from_str("event.params.asset.0").unwrap();
+        assert!(matches!(
+            arg,
+            CallArg::Ethereum(EthereumArg::ParamField(_, _))
+        ));
+        if let CallArg::Ethereum(EthereumArg::ParamField(param, field)) = arg {
+            assert_eq!(param.as_str(), "asset");
+            assert_eq!(field.as_str(), "0");
+        }
+    }
+
+    #[test]
+    fn test_struct_field_call_expr_parsing() {
+        // Test struct field access with numeric indices: ERC20[event.params.asset.1].name()
+        let expr: CallExpr = "ERC20[event.params.asset.1].name()".parse().unwrap();
+        assert_eq!(expr.abi, "ERC20");
+        assert_eq!(
+            expr.address,
+            CallArg::Ethereum(EthereumArg::ParamField("asset".into(), "1".into()))
+        );
+        assert_eq!(expr.func, "name");
+        assert_eq!(expr.args, vec![]);
+
+        // Test struct field access in arguments with numeric indices
+        let expr: CallExpr =
+            "Contract[event.address].transfer(event.params.data.0, event.params.data.1)"
+                .parse()
+                .unwrap();
+        assert_eq!(expr.abi, "Contract");
+        assert_eq!(expr.address, CallArg::Ethereum(EthereumArg::Address));
+        assert_eq!(expr.func, "transfer");
+        assert_eq!(
+            expr.args,
+            vec![
+                CallArg::Ethereum(EthereumArg::ParamField("data".into(), "0".into())),
+                CallArg::Ethereum(EthereumArg::ParamField("data".into(), "1".into()))
+            ]
+        );
+    }
+
+    #[test]
+    fn test_struct_field_access_functions() {
+        use ethabi::Token;
+
+        let tuple_fields = vec![
+            Token::Uint(ethabi::Uint::from(8u8)),     // index 0: uint8
+            Token::Address([1u8; 20].into()),         // index 1: address
+            Token::Uint(ethabi::Uint::from(1000u64)), // index 2: uint256
+        ];
+
+        // Test extract_struct_field with numeric indices
+        let struct_token = Token::Tuple(tuple_fields.clone());
+
+        // Test accessing index 0 (uint8)
+        let result = CallDecl::extract_struct_field(&struct_token, "0", "testCall").unwrap();
+        assert!(matches!(result, Token::Uint(_)));
```

Comment:
Better to compare with the actual `tuple_fields`, i.e. `assert_eq!(result, tuple_fields[0])` and similar for the other assertions